### ISSUE TITLE
Chaining .on() broken for null handlers in jQuery 2.2.0 (inconsistency with 1.12.0)

### ISSUE THREAD

**Issue by dwilson6** [jquery/jquery#2812]
It looks like a bug was introduced in jquery 2.2.0 with the 'on' function.

If you have code that registers a bunch of handlers in a chain and one of the callbacks is null, the on function will return the window object when the callback is null which causes the chaining to fail.  This works with jquery 2.1.4.  After a little investigation, it looks like the 'on' method was refactored to be global within jquery's scope and returns this (which used to be the jquery object but now is window).

//example from jquery impromptu plugin
$('.selector').on("keydown",keyDownEventHandler)
						.on('impromptu:loaded', opts.loaded)
						.on('impromptu:close', opts.close)
						.on('impromptu:statechanging', opts.statechanging)
						.on('impromptu:statechanged', opts.statechanged);

**Comment by dmethvin** [comment]
A callback can't be `null`, it's not a documented input. Previously that silently did nothing but it seems like that makes it harder to find the bug causing a null input.

**Comment by mgol** [comment]
It's weird, though, we brought the check for `null` back in `2.2.0`, there's a test for the `click` event:
https://github.com/jquery/jquery/blob/33b548c8e3d43b2ebdfb129fd8086a3b0c905919/test/unit/event.js#L8-L21.

**Comment by dwilson6** [comment]
Correct.  The callback doesn't necessarily have to be null but if its not a function, window is returned

**Comment by dwilson6** [comment]
After looking more at the source code, it seems like this reverted commit is the culprit (the on function was broken out so 'this' has a different meaning now): https://github.com/jquery/jquery/commit/3cc496ceecdbd76b6c556e85b12696f7c4c09dd5

**Comment by mgol** [comment]
A function or `false` which is treated specially. As @dmethvin mentioned, this is not a supported input so we can technically do what we want here but I'd like to see what's really happening here. I don't immediately see why the `this` leaks the global object since the functions in which it appears are supposed to be bound to the proper elements.

Thanks for the report!

**Comment by mgol** [comment]
@dwilson6 
> After looking more at the source code, it seems like this reverted commit is the culprit (the on function was broken out so 'this' has a different meaning now): 3cc496c

Good point, that's it!

Note that we've removed this check on `master`, using `.on(event, null)` will throw in jQuery 3.0.0. We've only restored it here so that both 1.12.0 & 2.2.0 don't throw... but the fix is buggy as it seems.

I'd fix it on the `2.2-stable` branch but this doesn't seem important enough to warrant a `2.2.1` release so unless we find something important to patch this will most likely stay as it is.

Thoughts, @jquery/core?

**Comment by dmethvin** [comment]
The code had been refactored as a non-method function to hide the private `one` arg, that's when `this` changed meanings. I don't think this is critical enough to force an immediate 2.2.1 either, especially since it's not documented.

If we have a unit test in there it seems like we're saying we want a non-`false` but falsy handler to be a chainable no-op. I really dislike that kind of behavior because it lets mistakes happen a lot easier, but we should decide what we want to do and make sure we're testing it right.

**Comment by herbalite** [comment]
The issue I encountered is somewhat similar to this here, as it relates to the event refactoring. If you think this warrants it's own feature/bug report that is fine with me.

Here's what happened:

After upgrading to jQuery 2.2.0 hoverintent for ui-tabs threw an error e.preventDefault() is not a function in the user defined event handler callback function, even though a quick test showed that typeof e.preventDefault returned 'function'. 

The code I was using is an updated version of this http://stackoverflow.com/questions/15346939/adding-hoverintent-to-jquery-ui-tabs.

The most important change from that code is the line

```javascript
jQuery.event.handle.apply( that, args );
```

```javascript
jQuery.event.dispatch.apply( that, args );
```
And i had changed from bind()/unbind() to on()/off(). I reverted to bin()/unbind() during bug triaging, but that made no difference to the error.

After unsuccessfully trying to pin point the problem I found this jquery-ui example

http://jqueryui.com/accordion/#hoverintent

Using that code my ui-tabs with hoverintent and hoverintent in general works again, although I did made some refactoring (using on()/off() instead of bind()/unbind(), and giving the events a namespace).

Since this issue shows up when using special event handlers, it might have an effect on other plug-ins, I hope this report helps to save people from spending a lot of time to fix similar issues.

Probably needs no fixing in jQuery, altough the fixed version is less performant than using jQuery.event.dispatch directly as it doesn't have to do all the work that is done in the trigger function (which seems to be a special event in itself)

**Comment by dwilson6** [comment]
Yeah, I agree that its probably not critical enough for a quick 2.2.1 release.  It could potentially break plugins that chain on like the one above where not all options are always set though so it might be worth fixing for compatibility.  Though plugins shouldn't be doing that anyways.

**Comment by mgol** [comment]
@dwilson6 Would you like to submit a patch against the `2.2-stable` branch? Judging by #2822 we might need to release 1.12.1/2.2.1 anyway so this is a good candidate to make the branches consistent again.

**Comment by mgol** [comment]
I've added a milestone.

**Comment by dwilson6** [comment]
Yeah, I will do that

**Comment by timmywil** [comment]
Fixed by https://github.com/jquery/jquery/issues/2825

**Comment by vankampenp** [comment]
The current 2.2.0 breaks jQuery Mobile selectMenu with data-native-menu="false". see http://plnkr.co/edit/dant044hERO4zpEJgB8S

**Comment by timmywil** [comment]
@vankampenp If it's related to this issue, it has already been fixed and will be released in 2.2.1.

**Comment by Gemorroj** [comment]
Not work fine, see https://github.com/jquery/jquery-mobile/issues/8381#issuecomment-187637179
